File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        lookup = {val:key for key, val in enumerate(inorder)}
        count = 0

        def rec(n, i, j):
            nonlocal count
            if n >= len(preorder):
                return None
            if i > j:
                return None
            count += 1
            root = TreeNode(preorder[n])
            index = lookup[preorder[n]]
            root.left = rec(count, i, index - 1)
            root.right = rec(count, index + 1, j)
            return root
        
        return rec(0, 0, len(inorder)-1)

        

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root, k) {
        let res = this.transversal(root, 0, k)
        return res[1]
        
    }

    transversal(root, count , k){
        if(root == null){
            return [0 + count, null]
        }
        console.log(root.val)
        let l = this.transversal(root.left, count ,k) 
        let m = [l[0] + 1, l[1]]
        console.log(m)
        let r = this.transversal(root.right, m[0], k) 
        if(l[1] != null){
            r[1] = l[1]
        }
        else if( m[0] == k){
            r[1] = root.val
        }
        return r 
    }
}
